### PR TITLE
Added Machine learning Workshop By Agrim

### PR DESCRIPTION
**Pull request by Agrim-Bansal**
In this workshop, the participants will make a simple machine learning model which predicts the relation between 2 sets of numbers.

The recording of workshop by me is at : https://hackclub.slack.com/archives/C01A9V5LPD0/p1600780766004400

### CODE REVIEW

File: workshops/intro-to-ML/README.md [hunk 1]
```diff
@@ -0,0 +1,1694 @@
...
+    1/1 [==============================] - 0s 1ms/step - loss: 2.0053e-07
+    Epoch 636/700
+    1/1 [==============================] - 0s 2ms/step - loss: 1.9669e-07
+    Epoch 637/700
+    1/1 [==============================] - 0s 2ms/step - loss: 1.9315e-07
+    Epoch 638/700
+    1/1 [==============================] - 0s 2ms/step - loss: 1.8945e-07
+    Epoch 639/700
+    1/1 [==============================] - 0s 2ms/step - loss: 1.8597e-07
+    Epoch 640/700
```
Comment (sampoder) [inline workshops/intro-to-ML/README.md:1489]: this makes the workshop super long, I'd remove it

File: workshops/intro-to-ML/README.md [hunk 2]
```diff
@@ -0,0 +1,1694 @@
+---
+name: 'Intro To ML'
+description: 'Making an Machine Learning model to predict relation between 2 number sets '
+author: '@Agrim-Bansal'
+---
+
```
Comment (sampoder) [inline workshops/intro-to-ML/README.md:6]: delete weird spacing

File: workshops/intro-to-ML/README.md [hunk 3]
```diff
@@ -0,0 +1,1694 @@
...
+
+
+
+
+# Introduction
+
+
+Today We'll make a machine learning model where it learns the relationship between two sets of numbers.
```
Comment (sampoder) [inline workshops/intro-to-ML/README.md:14]: let's make this a bit more exciting

File: workshops/intro-to-ML/README.md [hunk 4]
```diff
@@ -0,0 +1,1694 @@
...
+
+<br/>
```
Comment (sampoder) [inline workshops/intro-to-ML/README.md:16]:
```suggestion
```

File: workshops/intro-to-ML/README.md [hunk 5]
```diff
@@ -0,0 +1,1694 @@
...
+
+# Machine Learning
+Machine learning is very similar to Artificial Intelligence but it is not exactly it. Machine learnig (or ML as it is called) is all about data and classification.
+
+When we as programmers code a normal program, there we define rules and input some test data and at end, obtain results.
+
+But in machine learning, the process is somewhat different. Here we provide the machine some input data and some output data ie the known test data and it is the task of the machine to figure out what is the relation between the sets of data and further predict output values for unknown input data
+
+Lets come to the process
```
Comment (sampoder) [inline workshops/intro-to-ML/README.md:25]:
```suggestion
**How does it work?**
```

File: workshops/intro-to-ML/README.md [hunk 6]
```diff
@@ -0,0 +1,1694 @@
...
+We provide the machine with some sets of data (usually 2) and it is made to establish a link between the data sets or "the ML model is trained". 
+In this process of training, the machine makes a guess and then checks how good or bad the guess was and it calculates the loss. Then it uses an optimzer to make another optimized guess which is better than the previous one or the loss is decreased. This process continues as many times as we specify.
```
Comment (sampoder) [inline workshops/intro-to-ML/README.md:27]: maybe use prediction instead of guess

File: workshops/intro-to-ML/README.md [hunk 7]
```diff
@@ -0,0 +1,1694 @@
...
+But how can we train a comuter to do the equivalent task i.e. to figure out the relationship and predict for a future number?
+Using data! 
+
+We'll be feeding it with a set of Xs, and a set of Ys, and when we train it then it should be able to figure out the relationship between them. 
+
+So, lets get started :)
+
+Note : for this project We use PYTHON and a library called Tensorflow
+
+<br/>
```
Comment (sampoder) [inline workshops/intro-to-ML/README.md:46]: delete this weird spacing

File: workshops/intro-to-ML/README.md [hunk 8]
```diff
@@ -0,0 +1,1694 @@
...
+<br/>
+
+
+
+
+# Using Google Colab
+
+We'll be making our project on a google colab notebook.
+Go to https://colab.research.google.com and sign in with your google account.
```
Comment (sampoder) [inline workshops/intro-to-ML/README.md:58]:
```suggestion
Go to https://colab.research.google.com and sign in with your Google account.
```

File: workshops/intro-to-ML/README.md [hunk 9]
```diff
@@ -0,0 +1,1694 @@
...
+
+Then  Click on File > New notebook
+
+<br />
+
+Connect to a machine
+
+<br />
```
Comment (sampoder) [inline workshops/intro-to-ML/README.md:66]: delete spacing

File: workshops/intro-to-ML/README.md [hunk 10]
```diff
@@ -0,0 +1,1694 @@
...
+
+Add a code block and we are ready to get started !
+
+(Press ctrl + Enter to run the code)
```
Comment (sampoder) [inline workshops/intro-to-ML/README.md:70]:
```suggestion
(Press <kbd>CTRL</kbd> + <kbd>Enter</kbd> to run the code)
```

File: workshops/intro-to-ML/README.md [hunk 11]
```diff
@@ -0,0 +1,1694 @@
...
+
+<br/>
+
+<br/>
```
Comment (sampoder) [inline workshops/intro-to-ML/README.md:74]: pls delete

File: workshops/intro-to-ML/README.md [hunk 12]
```diff
@@ -0,0 +1,1694 @@
...
+
+
+# Imports
+
+Let's start with our imports. Imports provide us a way to exceed the limits of inbuilt modules and functions.
+They are just some code peices written by developers which make repetitive tasks way more easier and help in reducing out code
```
Comment (sampoder) [inline workshops/intro-to-ML/README.md:80]: spelling

File: workshops/intro-to-ML/README.md [hunk 13]
```diff
@@ -0,0 +1,1694 @@
...
+
+To import a module (called a library) the syntax is :
+```import <library_name> ```
+
+
+<br />
```
Comment (sampoder) [inline workshops/intro-to-ML/README.md:86]: weird spacing delete

File: workshops/intro-to-ML/README.md [hunk 14]
```diff
@@ -0,0 +1,1694 @@
...
+<br/>
+
+
+# Imports
+
+Let's start with our imports. Imports provide us a way to exceed the limits of inbuilt modules and functions.
+They are just some code peices written by developers which make repetitive tasks way more easier and help in reducing out code
+
+To import a module (called a library) the syntax is :
+```import <library_name> ```
```
Comment (sampoder) [inline workshops/intro-to-ML/README.md:83]:
```suggestion
```
import <library_name> 
```
```

File: workshops/intro-to-ML/README.md [hunk 15]
```diff
@@ -0,0 +1,1694 @@
...
+
+
+<br />
+
+
+Further, we can import something specific instead of the  complete library(to reduce program size) from the library using :
+
+```from <library_name> import <object_name>```
```
Comment (sampoder) [inline workshops/intro-to-ML/README.md:91]: see above for single line syntax

File: workshops/intro-to-ML/README.md [hunk 16]
```diff
@@ -0,0 +1,1694 @@
...
+
+
+<br />
+
+
+Further, we can import something specific instead of the  complete library(to reduce program size) from the library using :
+
+```from <library_name> import <object_name>``` 
+
+<br />
```
Comment (sampoder) [inline workshops/intro-to-ML/README.md:93]: delete

File: workshops/intro-to-ML/README.md [hunk 17]
```diff
@@ -0,0 +1,1694 @@
...
+
+
+```
+import tensorflow as tf
+import numpy as np
+from tensorflow import keras
+```
+
+<br/>
+<br/>
```
Comment (sampoder) [inline workshops/intro-to-ML/README.md:133]: ... delete

File: workshops/intro-to-ML/README.md [hunk 18]
```diff
@@ -0,0 +1,1694 @@
...
+
+![neural network](https://cloud-ltmlw2jd1.vercel.app/neural.png)
+
+```
+model = tf.keras.Sequential([keras.layers.Dense(units=1, input_shape=[1])])
+```
+
+Here, model is a variable which stores the neural network to be referred further in the program.
+
+<br/><br/><br/>
```
Comment (sampoder) [inline workshops/intro-to-ML/README.md:151]: .....

File: workshops/intro-to-ML/README.md [hunk 19]
```diff
@@ -0,0 +1,1694 @@
...
+Over time we will learn the different and appropriate loss and optimizer functions for different scenarios. 
+
+
+
+```
+model.compile(optimizer='sgd', loss='mean_squared_error')
+```
+
+
+<br/><br/><br/>
```
Comment (sampoder) [inline workshops/intro-to-ML/README.md:176]: ... delete

File: workshops/intro-to-ML/README.md [hunk 20]
```diff
@@ -0,0 +1,1694 @@
...
+
+# Training the Neural Network
+
+The process of training the neural network , where it 'learns' the relationship between the Xs and Ys is in the **model.fit**  call. This is where it will go through the loop we spoke about above, making a guess, measuring how good or bad it is (i.e. the loss), using the opimizer to make another guess etc. It will do it for the number of epochs we specify. When we run this code, we'll see the loss on the right hand side.
+
+
+```
+model.fit(xs, ys, epochs=700)
+```
+<style>
```
Comment (sampoder) [inline workshops/intro-to-ML/README.md:203]: won't render

**Review by sampoder — CHANGES_REQUESTED** [review]
Hi! This workshop is super close to being ready I sense, needs a few changes.

- Please check the formatting of this .md file, the spacing is super weird (no `<br />` please)
- Please add 3 demos of where to go beyond
- Please check the grammar of this workshop.

Thanks 🚀

**Comment by Agrim-Bansal** [comment]
Surely I'll make all the changes.
Thank you so much for your guidance

Agrim

File: workshops/intro-to-ML/README.md (hunk 3 above)
Comment (Agrim-Bansal) [reply workshops/intro-to-ML/README.md:14]: Could You please give some suggestions for it

Comment (sampoder) [reply workshops/intro-to-ML/README.md:8]: take a look at some others on the site to get an idea, sorry im busy rn

**Comment by Agrim-Bansal** [comment]
I've made all the corrections. Please have a look and let me know if anything else is required

**Comment by sampoder** [comment]
Could we make this workshop easier to read?

For example paragraphs should have spacing in-between paragraphs and add code highlighting

**Comment by sampoder** [comment]
From formatting perspective

**Review by sampoder — CHANGES_REQUESTED** [review]
View above

File: workshops/intro_to_ML/README.md [hunk 21]
```diff
@@ -0,0 +1,220 @@
...
+
+Here we are importing the library ["TensorFlow"](https://www.tensorflow.org/api_docs) and calling it "tf" for ease of use.
+This library contains many tools for machine learning   
+We then import a library called ['Numpy'](https://numpy.org/doc/), which helps us to represent and manipulate our data as lists easily and quickly. We call it as "np"for ease.
+It provides a quicker and easier way to manipulate lists as compared to python's inbuilt listing system.
+Numpy documentation : 
+
+The framework for defining a neural network as a set of Sequential layers is called "keras", so we import that too.We're importing it separately for our ease
+
+```
```
Comment (faisalsayed10) [inline workshops/intro_to_ML/README.md:65]: Add syntax highlighting everywhere after the 3 backticks
eg: for javascript it should be 
`(``` js)`
and similarly for HTML and CSS
`(```html)`
`(``` css)`

Comment (faisalsayed10) [inline workshops/intro_to_ML/README.md:65]:
```suggestion
```js
```

**Review by faisalsayed10 — CHANGES_REQUESTED** [review]

File: workshops/intro_to_ML/README.md (hunk 21 above)
Comment (sampoder) [reply workshops/intro_to_ML/README.md:65]: @faisalsayed10 this python code lol, not js

**Comment by Agrim-Bansal** [comment]
I've added the syntax highlighting.
Should I add the spacing in the code for the markdown file or should it reflect in the preview ?

**Comment by Agrim-Bansal** [comment]
> Could we make this workshop easier to read?
> 
> For example paragraphs should have spacing in-between paragraphs and add code highlighting

I've done these things.
Please have a look at it and let me know what else is required @sampoder

File: workshops/intro_to_ML/README.md (hunk 21 above)
Comment (faisalsayed10) [reply workshops/intro_to_ML/README.md]: oof

File: workshops/intro_to_ML/README.md [hunk 22]
```diff
@@ -0,0 +1,272 @@
+---
+name: 'Intro To ML'
+description: 'Making a Machine Learning model to predict relation between 2 number sets '
+author: '@Agrim-Bansal'
+---
+
+
+# Introduction
```
Comment (sampoder) [inline workshops/intro_to_ML/README.md:8]: please make this the title of the workshop

File: workshops/intro_to_ML/README.md [hunk 23]
```diff
@@ -0,0 +1,272 @@
...
+
+
+You might have thought 21, right? But it ended up being a little over. Why do you think that is?  
+
+Remember that the neural networks deal with probabilities, so given the data that we fed the Neural network with, it calculated that there is a very high probability that the relationship between X and Y is Y=2x+1, but with only 6 data values we can't know for sure. As a result, the value for 10 is very close to 21, but not necessarily 21.  
+
+As We work with neural networks, we'll see this pattern recurring. We will almost always deal with probabilities, not certainties, and will do a little bit of coding to figure out what the result is based on the probabilities, particularly when it comes to classification.  
+
+
+<br/>
```
Comment (sampoder) [inline workshops/intro_to_ML/README.md:230]: what's with all this spacing?

File: workshops/intro_to_ML/README.md [hunk 24]
```diff
@@ -0,0 +1,272 @@
...
+
+```py
+model.fit(xs, ys, epochs=700)
+```
+
+```py
+    Epoch 1/700
+    1/1 [==============================] - 0s 1ms/step - loss: 8.0000
+    Epoch 2/700
+    1/1 [==============================] - 0s 2ms/step - loss: 5.5120
```
Comment (sampoder) [inline workshops/intro_to_ML/README.md:159]: this is still massive and not really needed

Comment (Agrim-Bansal) [reply workshops/intro_to_ML/README.md:159]: I added it so that the participants have an idea that what they should see after running the code.
Shall I remove it?

File: workshops/intro_to_ML/README.md (hunk 23 above)
Comment (Agrim-Bansal) [reply workshops/intro_to_ML/README.md:230]: It does not reflect in the output. I just added it to make it more readable

File: workshops/intro_to_ML/README.md (hunk 22 above)
Comment (Agrim-Bansal) [reply workshops/intro_to_ML/README.md:8]: Could you please elaborate a bit? I couldn't understand

Comment (Agrim-Bansal) [reply workshops/intro_to_ML/README.md:8]: @sampoder

File: workshops/intro_to_ML/README.md (hunk 24 above)
Comment (sampoder) [reply workshops/intro_to_ML/README.md:159]: maybe a screenshot

File: workshops/intro_to_ML/README.md (hunk 23 above)
Comment (sampoder) [reply workshops/intro_to_ML/README.md:230]: can you remove it please, it just looks off

File: workshops/intro_to_ML/README.md (hunk 22 above)
Comment (sampoder) [reply workshops/intro_to_ML/README.md:8]: just make it # Intro to ML

Comment (sampoder) [reply workshops/intro_to_ML/README.md:8]: just so that the site works

**Comment by Agrim-Bansal** [comment]
@sampoder, I've done all the changes. Are they fine ?

**Comment by sampoder** [comment]
looks good to me, @faisalsayed10 thoughts?

**Comment by faisalsayed10** [comment]
taking a look into it!

File: workshops/intro_to_ML/README.md [hunk 25]
```diff
@@ -0,0 +1,213 @@
...
+```
+
+We can easily figure out the relation between the 2 given sets which is 2x + 1.   
+But how can we train a computer to do the equivalent task i.e. to figure out the relationship and predict for a future number?  
+Using data!   
+We'll be feeding it with a set of Xs and a set of Ys, and when we train it then it should be able to figure out the relationship between them.  
+
+So, lets get started :)  
+<br/>
+Note : for this project We'll use PYTHON and a library called Tensorflow.
```
Comment (faisalsayed10) [inline workshops/intro_to_ML/README.md:42]:
```suggestion
Note : for this project We'll use Python and a library called Tensorflow.  
```

**Review by faisalsayed10 — APPROVED** [review]
Just this one small thing and everything is good!

**Comment by faisalsayed10** [comment]
Looks good to me too!

**Comment by Hugoyhu** [comment]
Hi, could you please provide a slack handle and the name of the Hack Club you lead? thanks!

**Comment by Agrim-Bansal** [comment]
Hi, I currently don't lead a hackclub at my school but I'm planning to run it once I can persuade the teachers to allow us.
And I'm 'Agrim' on slack

File: workshops/intro_to_ML/README.md (hunk 25 above)
Comment (Agrim-Bansal) [reply workshops/intro_to_ML/README.md]: I've made the changes.
Thank you

**Comment by Hugoyhu** [comment]
Are you an active slack member?

On Oct 7, 2020, at 11:29 PM, Agrim Bansal <notifications@github.com<mailto:notifications@github.com>> wrote:



Hi, I currently don't lead a hackclub at my school but I'm planning to run it once I can persuade the teachers to allow us.
And I'm 'Agrim' on slack

**Comment by Agrim-Bansal** [comment]
I participate in almost all of the events (except those with the time zone problem). 
I also take part in the discussions where I have the knowledge to be involved. You can find me in some discussions in  #india, #hardware, #code, #hardware-party, etc channels. And I'm definitely **not** leaving hack club

**Review by lachlanjc — CHANGES_REQUESTED** [review]
Hey Agrim, thanks for your submission! This is a good start for a workshop, but you should focus on what you make/what you can do with it, not the topic of the tutorial. It’s not clear what you can do with this, or why you would want to follow it—you can use the same code, but it should be used to make something visual or real-world applicable, even in a simple way, instead of just numbers.

As far as the specifics/formatting of the workshop itself:
- Can you use a more descriptive name? We already have some ML intro workshops & you want to focus on the output
- The punctuation + grammar of this workshop have a lot of mistakes, & are very inconsistent with spaces, etc
- Can you add some more images/make it more visual? The Compilation, Prediction etc steps are hard to follow
- Please remove all the `<br>` tags
- Replace all the `# ` with `## `

We’d love to have another ML workshop, so if you’d like to keep working on this & address these issues, go for it! But currently this doesn’t meet our guidelines unfortunately.
